pallets/xtransfer/src/xcm/xcm_transfer.rs

```diff
@@ -154,17 +159,26 @@ pub mod pallet {
 			T::Weigher::weight(&mut msg).map_or(Weight::max_value(), |w| w)
 		}
 
-		pub fn do_transfer(
-			origin: OriginFor<T>,
+		pub fn do_transfer_multiasset(
+			sender: T::AccountId,
+			origin_location: MultiLocation,
 			asset: MultiAsset,
-			para_id: ParaId,
-			recipient: T::AccountId,
-			amount: BalanceOf<T>,
+			dest: MultiLocation,
 			dest_weight: Weight,
 		) -> DispatchResult {
-			let sender = ensure_signed(origin.clone())?;
-			let origin_location = T::ExecuteXcmOrigin::ensure_origin(origin)?;
-			let dest_location = (1, X1(Parachain(para_id.into()))).into();
+			let mut dest_location = dest.clone();
+			// make sure we are processing crosschain transfer and we got correct path
+			ensure!(!dest_location.is_here(), Error::<T>::IllegalDestination);
+			// FIXME: what if someone give a Parachain junction at the end?
+			// After take_last(), dest only contains reserve location of the recipient.
+			let recipient = match dest_location.take_last() {
+				Some(Junction::AccountId32 {
+					network: _,
+					id: recipient,
+				}) => Some(recipient),
+				_ => None,
+			};
```

Comment:
Recipient was taken from the dest_location, but after that we still construct it back as a MultiLocation. Is it possible to pass the recipient directly to the session to avoid double conversion?

https://github.com/Phala-Network/khala-parachain/blob/0e5f8ec317f08cdb874d74f78ca69563693ab019/pallets/xtransfer/src/xcm/xcm_transfer.rs#L367-L371

Reply:
We'd better extract it from dest, as the dest is passed from our client or evm chain. Otherwise we need a extra parmeter.

Reply:
Not quite understand. Let's have an offline chat about this